### PR TITLE
Implement NBT/Metadata Serializer

### PR DESCRIPTION
**Pull request by geNAZt**
Added Metadata serializer
Refactored ItemData / BlockData to one generic Metadata API
Implemented most Item Metadata

### CODE REVIEW

File: api/src/main/java/com/voxelwind/api/game/level/block/data/Cake.java [hunk 1]
```diff
@@ -1,20 +1,18 @@
...
  */
-public class Cake implements BlockData {
+public class Cake implements Metadata {
     private final int level;
 
     public static final Cake NEW = new Cake(0);
     public static final Cake ALMOST_EATEN = new Cake(6);
 
-    public static Cake ofStage(int data) {
-        Preconditions.checkArgument(data >= 0 && data < 7, "data is not valid (wanted 0-6)");
```
Comment (minecrafter) [inline api/src/main/java/com/voxelwind/api/game/level/block/data/Cake.java]: Can you explain why you removed this validation check?

File: api/src/main/java/com/voxelwind/api/game/level/block/data/Crops.java [hunk 2]
```diff
@@ -1,21 +1,19 @@
...
  */
-public class Crops implements BlockData {
+public class Crops implements Metadata {
     private final int level;
 
     public static final Crops NEW = new Crops(0);
     public static final Crops FULLY_GROWN = new Crops(7);
 
-    public static Crops ofStage(int data) {
-        Preconditions.checkArgument(data >= 0 && data < 8, "data is not valid (wanted 0-7)");
```
Comment (minecrafter) [inline api/src/main/java/com/voxelwind/api/game/level/block/data/Crops.java]: Can you explain why you removed this validation check?

**Comment by minecrafter** [comment]
LGTM once validation is added back to API functions

File: api/src/main/java/com/voxelwind/api/game/level/block/data/Cake.java (hunk 1 above)
Comment (geNAZt) [inline api/src/main/java/com/voxelwind/api/game/level/block/data/Cake.java]: The validation should be done in the generation part. The idea behind the of() static helpers is to keep them until they get replaced by a Builder. I simply missed the // TODO here

**Pull request merged**